cornac/models/nmf/recom_nmf.pyx

```diff
@@ -0,0 +1,275 @@
+            long num_items = self.train_set.num_items
+            long num_ratings = val.shape[0]
+            int num_factors = self.k
+            int max_iter = self.max_iter
+            int num_threads = self.num_threads
+            floating lr = self.learning_rate
+
+            floating lambda_u = self.lambda_u
+            floating lambda_v = self.lambda_v
+            floating lambda_bu = self.lambda_bu
+            floating lambda_bi = self.lambda_bi
+            floating mu = self.global_mean
+
+            bool use_bias = self.use_bias
+            bool verbose = self.verbose
+
+            floating[:, :] U_numerator = np.zeros((num_users, num_factors), dtype=np.float32)
+            floating[:, :] V_numerator = np.zeros((num_items, num_factors), dtype=np.float32)
+            floating[:, :] U_denominator = np.zeros((num_users, num_factors), dtype=np.float32)
+            floating[:, :] V_denominator = np.zeros((num_items, num_factors), dtype=np.float32)
+
+            floating loss, r, r_pred, error, eps = 1e-9
+
+            long u, i, f, j
+
+        from tqdm import trange
+        progress = trange(max_iter, disable=not verbose)
+        for epoch in progress:
+            loss = 0.
+
+            for j in prange(num_ratings, nogil=True, num_threads=num_threads):
+                u, i, r = rid[j], cid[j], val[j]
+
+                # predict rating
+                r_pred = mu + Bu[u] + Bi[i]
+                for f in range(num_factors):
+                    r_pred = r_pred + U[u, f] * V[i, f]
+
+                error = r - r_pred
+                loss += error * error
+
+                # update biases
+                if use_bias:
+                    Bu[u] += lr * (error - lambda_bu * Bu[u])
+                    Bi[i] += lr * (error - lambda_bi * Bi[i])
+
+                # compute numerators and denominators
+                for f in range(num_factors):
+                    U_numerator[u, f] += r * V[i, f]
+                    U_denominator[u, f] += r_pred * V[i, f]
+                    V_numerator[i, f] += r * U[u, f]
+                    V_denominator[i, f] += r_pred * U[u, f]
+
+            # update user factors
+            for u in prange(num_users, nogil=True, num_threads=num_threads):
+                for f in range(num_factors):
+                    loss += lambda_u * U[u, f] * U[u, f]
+                    U_denominator[u, f] += user_counts[u] * lambda_u * U[u, f] + eps
+                    U[u, f] *= U_numerator[u, f] / U_denominator[u, f]
+
+            # update item factors
+            for i in prange(num_items, nogil=True, num_threads=num_threads):
+                for f in range(num_factors):
+                    loss += lambda_v * V[i, f] * V[i, f]
+                    V_denominator[i, f] += item_counts[i] * lambda_v * V[i, f] + eps
+                    V[i, f] *= V_numerator[i, f] / V_denominator[i, f]
+
+            # reset all numerators and denominators
```

Comment:
Is there a better way to reset the values to zero rather than allocating new mem?
For example https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.fill.html

Is it more intuitive to set the values to zero before the loops rather than putting it here?